lib/renderers/data_table_csv_renderer.rb

```diff
@@ -1,27 +1,88 @@
-      filename = options[:filename]
-    elsif model.is_a?(DataTable::Base)
-      filename = model.download_file_name(self)
-    else
-      raise ArgumentError "Invalid model: #{model.class}"
-    end
-
-    options[:columns_layout] = UserGridLayout.get_user_grid_layout(current_user, self.controller_name, self.action_name, model.class.name, model.additional_layout_key)
-
-    new_params = params.merge(
-      'filterOptions' => options.dig(:columns_layout, 'filterOptions'),
-      'sortOptions' => options.dig(:columns_layout, 'sortOptions')
-    )
-
-    send_data(
-      model.to_csv(view_context, new_params, options),
-      :type => mime_type,
-      :disposition => "attachment; filename=#{filename}.#{extension}"
-    )
-  end
-end
+require 'csv'
+module Devextreme
+  ActionController::Renderers.add :data_table_csv do |model, options|
+    extension   = options[:extension] || 'csv'
+    mime_type   = options[:mime_type] || Mime[:csv]
+    if options[:filename].present?
+      filename = options[:filename]
+    elsif model.is_a?(DataTable::Base)
+      filename = model.download_file_name(self)
+    else
+      raise ArgumentError "Invalid model: #{model.class}"
+    end
+
+    options[:columns_layout] = UserGridLayout.get_user_grid_layout(current_user, self.controller_name, self.action_name, model.class.name, model.additional_layout_key)
+
+    new_params = params.merge(
+      'filterOptions' => options.dig(:columns_layout, 'filterOptions'),
+      'sortOptions' => options.dig(:columns_layout, 'sortOptions')
+    )
+
+    send_data(
+      model.to_csv(view_context, new_params, options),
+      :type => mime_type,
+      :disposition => "attachment; filename=#{filename}.#{extension}"
+    )
+  end
+
+  class DataTableCsvGenerator
+    def initialize(data_table, view_context, query, options = {})
+      @data_table = data_table
+      @view_context = view_context
+      @query = query
+      @options = options
+      @summary = @options.delete(:summary)
+    end
+
+    def run
+      generate_csv
+    end
+
+    private
+
+    def generate_csv
+      csv_rows = []
+
+      if @summary && @summary.count > 0
+        @summary.each do |row|
+          csv_rows << row
+        end
+        csv_rows << []
+      end
+
+      columns = @options.fetch(:columns_layout, {}).fetch('columns', [])
+
+      if columns.present?
+        @data_table.columns.each do |column|
+          user_column = columns.detect { |c| c['dataField'].split('.').last == column.name.to_s } || { 'visible' => false }
+          column.options.reverse_merge!(user_visible: user_column['visible'], user_visible_index: user_column['visibleIndex'])
+        end
+      end
+
+      unless @options.has_key?(:write_headers) && !@options[:write_headers]
+        headers = @data_table.each_header
+        csv_rows << headers
+      end
+
+      @query.each do |instance|
+        row = []
+        @data_table.each_row(instance, @view_context).each do |value|

```

Comment:
the logic for `each_row` will have to change.

it currently calls `to_xls_text`.

we will have to enhance/change it to call `to_csv_text` when generating csv's

possible implementation:
```
def each_xls_row(instance, view_context)
  each_row(instance, view_context, :to_xls_text)
end

def each_csv_row(instance, view_context)
  each_row(instance, view_context, :to_csv_text)
end

private def each_row(instance, view_context, method_call)
  @columns
    .select{ |column| column.options.fetch(:user_visible, true) == true }
    .sort_by{ |column| column.options[:user_visible_index] }
    .collect{ |column| column.send(:method_call, instance, view_context) rescue nil }
end
```

then change `DataTableXlsGenerator` to use the new method for `each_xls_row`